src/btree/bt_cursor.c

```diff
@@ -978,20 +1084,31 @@ __cursor_truncate(WT_SESSION_IMPL *session,
 	 * instantiated the end cursor, so we know that page is pinned in memory
 	 * and we can proceed without concern.
 	 */
-retry:	WT_RET(__wt_btcur_remove(start));
+retry:	WT_RET(__wt_btcur_search(start));
+
+	/*
+	 * XXX KEITH
+	 * When the btree cursor code sets/clears the cursor flags (rather than
+	 * the cursor layer), the set/clear goes away, only the assert remains.
+	 */
+	F_CLR((WT_CURSOR *)start, WT_CURSTD_KEY_SET);
+	F_SET((WT_CURSOR *)start, WT_CURSTD_KEY_INT);
+	WT_ASSERT(session,
+	    F_MASK((WT_CURSOR *)start, WT_CURSTD_KEY_SET) == WT_CURSTD_KEY_INT);
 
 	/*
 	 * Reset ret each time through so that we don't loop forever in
 	 * the cursor equals case.
 	 */
 	for (ret = 0;;) {
+		if ((ret = rmfunc(session, start, 1)) != 0)
```

Comment:
Unfortunately, this change is necessary. Here's the deal:

* The truncate loop consists of repeated cursor `remove/next` call pairs.
* If `next` is called with an unpositioned cursor, it will move to the start of the object.
* One of the changes in this PR is that `remove` only returns with the cursor positioned if it starts with the cursor positioned. For `remove` to return with a positioned cursor (so we call `next` with a positioned cursor), then `remove` must start with a positioned cursor.

To ensure we always start the truncate loop with a positioned cursor, I think we can either ensure the session layer never calls truncate with an unpositioned cursor, or we can call search on the cursor passed into the truncate layer.

Happy to undo this change if there's a better approach.